staking/contracts/contracts.py

```diff
@@ -104,16 +115,20 @@ def calculate_rewards(addr: Expr) -> Expr:
         # Duration
         (duration := ScratchVar()).store(end.load() - start.load()),
 
-        # Calculate time since last updated
+        # Calculate rewards
         (rewards := ScratchVar()).store(
-            App.localGet(addr, Bytes("AS")) * duration.load() / Int(31557600) * App.globalGet(Bytes("FR")) / Int(10000)
+            App.localGet(addr, Bytes("AS")) * duration.load() / Int(31557600) * App.globalGet(Bytes("FR")) / Int(10000) # I think there might be truncation issues, especially with very high denominator values.
         ),
 
         # Remove rewards from global
-        App.globalPut(Bytes("TR"), App.globalGet(Bytes("TR")) - rewards.load()),
+        App.globalPut(Bytes("TR"), App.globalGet(Bytes("TR")) - rewards.load()), # What happens after "TR" runs out and goes to zero? I think the transaction will fail and any method (i.e. deposit, withdraw) that wants to calculate rewards will also fail. Would anybody be able to take out assets if this happens?
```

Comment:
I think subtracting `TR` should be done atomically with the `withdraw` operation, and should fail when trying to withdraw too many assets? 
Or alternatively, maybe we need a `TR>0` check somewhere earlier and independent of calculating rewards